### PR TITLE
implement growable arrays

### PR DESCRIPTION
**Pull request by rfomin**
Growable arrays of homogeneous values of any type, similar to a std::vector in C++. The array must be initialised to NULL. Inspired by https://github.com/nothings/stb/blob/master/stb_ds.h

We can implement this as a header only, but I think it's more convenient to encapsulate the inner structure.

### CODE REVIEW

File: src/m_array.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+struct buffer_s
+{
+    unsigned int capacity;
+    unsigned int size;
+    char buffer[];
+};
+
+#define ARR_PTR(v) \
+    ((struct buffer_s *)((char *)(v) - offsetof(struct buffer_s, buffer)))
```
Comment (rfomin) [inline src/m_array.c:28]: The trick is to use the `offsetof` macro from `stddef.h` and store the capacity and size in front of the array buffer.

Comment (fabiangreffrath) [reply src/m_array.c:28]: Never heard of this before, but it's brilliant!

File: src/m_array.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// and any previously-taken pointers should be considered invalidated.
+
+void *M_ArrayIncreaseCapacity(void *v, size_t esize, size_t n);
+void *M_ArrayIncreaseSize(void *v, size_t esize);
+
+size_t array_size(void *v);
+void   array_clear(void *v);
+void   array_free(void *v);
+
+#define array_grow(v, n) \
```
Comment (fabiangreffrath) [inline src/m_array.h:28]: Do we even need `array_grow`?

Comment (fabiangreffrath) [reply src/m_array.h:28]: I mean as an exported function.

Comment (rfomin) [reply src/m_array.h:28]: I wanted to use it in `dsdhacked.c`, but there are already a lot of changes.

Comment (fabiangreffrath) [reply src/m_array.h:28]: For everything related to e.g. keeping track of files we probably don't need more than `8` for a first guess. But bumping it to e.g. `32` shouldn't hurt either.

**Review by fabiangreffrath — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
In theroy, we could even replace `z_zone.c` with this approach, i.e. one growing list of pointers for each `PU_*` class. Not that I have such plans, but we *could*. :grin:

**Comment by rfomin** [comment]
> In theroy, we could even replace `z_zone.c` with this approach, i.e. one growing list of pointers for each `PU_*` class. Not that I have such plans, but we _could_. 😁

I think we should use it only for trivial code, not in renderer/playsim. Otherwise we should implement it as header only with more macros for optimal performance. And even then it will still be far from a real `std::vector`, e.g. no overflow checking, pointers will be invalidated because of `realloc`, etc.

**Pull request merged**